### PR TITLE
Fix a problem when a connection is rejected when the last handshake was missed

### PR DESCRIPTION
**Pull request by ethouris**
In the caller-listener configuration, the socket that is currently in the handshake (connecting) state expects only handshake messages to be received.

In a situation that the conclusion handshake was sent from a caller to a listener, the listener interprets it and turns into a connected state, with sending the last conclusion handshake response to the caller, then it states the caller is now connected and the application gets the accepted socket and can use it for sending data. However, if this last handshake was lost, the socket on the caller side is still in connecting state and receiving data packets (or anything else than handshake) makes it confused. The 1.3.0 version introduced a bug that in a situation like this makes the connection rejected.

The fix is to recognize this situation properly and make the caller repeat sending the last conclusion handshake, until the conclusion handshake response is finally received from the listener. Of course, any packets received from the listener would have to be ignored and the listener needs to repeat them.

This fix also contains a possibility to forcefully simulate this particular packet to be lost by defining the `TEST_DISABLE_FINAL_HANDSHAKE` macro (also through `./configure --srt-extra-cflags=-DTEST_DISABLE_FINAL_HANDSHAKE`).

### CODE REVIEW

**Comment by ethouris** [comment]
BLOCKED BY: #504

**Comment by ethouris** [comment]
BLOCKING #504 merged.

File: srtcore/core.cpp [hunk 1]
```diff
@@ -4604,7 +4625,10 @@ void CUDT::acceptAndRespond(const sockaddr* peer, CHandShake* hs, const CPacket&
            << "), target_socket=" << response.m_iID << ", my_socket=" << debughs.m_iID);
    }
 #endif
+
+#ifndef TEST_DISABLE_FINAL_HANDSHAKE
    m_pSndQueue->sendto(peer, response);
+#endif
```
Comment (maxlovic) [inline srtcore/core.cpp]: What is a purpose of this `TEST_DISABLE_FINAL_HANDSHAKE`?

File: srtcore/core.cpp [hunk 2]
```diff
@@ -3607,13 +3609,22 @@ EConnectStatus CUDT::processConnectResponse(const CPacket& response, CUDTExcepti
...
-       return CONN_REJECT;
+       if (!response.isControl())
+       {
+           LOGC(mglog.Error, log << CONID() << "processConnectResponse: received DATA while HANDSHAKE expected");
+       }
+       else
+       {
+           LOGC(mglog.Error, log << CONID()
+                   << "processConnectResponse: received NOT UMSG_HANDSHAKE before connection established: "
+                   << MessageTypeStr(response.getType(), response.getExtendedType()));
```
Comment (maxlovic) [inline srtcore/core.cpp]: This message is a bit confusing and rough for me. Maybe something like:
```
<< "processConnectResponse: expected HANDSHAKE before connection established, received: " 
<< MessageTypeStr(response.getType(), response.getExtendedType()));
```

**Pull request merged**